Replace linkage in WSG model with independent fingers.

avalenzu opened the pull request:
Coordination is now achieved by the controller. This results in ~2x
speedup for simulations that involve the WSG.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/robotlocomotion/drake/8819)
<!-- Reviewable:end -->

avalenzu commented:
+@sammy-tri for feature review.

+(status: curate commits before merging) 

---

Review status: 0 of 16 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LCak7HECz8ymZ7YTxYS:b7kv1hi)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
BTW I'm currently pausing on the review until CI passes.

---

Review status: 9 of 16 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LCoP4HO21S2spfoaFml:b42uqsn)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
-(status: curate commits before merging) . `bazel test //...` passes on my local machine. Should be good to continue reviewing now.

---

Review status: 4 of 16 files reviewed at latest revision, 2 unresolved discussions.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.h, line 18 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LCikPlP5tt86SQJEdOx:-LDJ2PH_2ftoIwE82Ech:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/e23ede3fe4cfc656a0fdf1a4db387500e10661b9/manipulation/schunk_wsg/schunk_wsg_plain_controller.h#L18)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

BTW I think this would likely be clearer with more documentation of the values/units/format/etc of the input and output ports (either in the class description or as comments on the accessors?)
</blockquote></details>

Done.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.h, line 60 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LCijWde3zX5I7CZdlWz:-LDJ2Pic10bPjPbnA4nX:bs7wuor) ([raw file](https://github.com/robotlocomotion/drake/blob/e23ede3fe4cfc656a0fdf1a4db387500e10661b9/manipulation/schunk_wsg/schunk_wsg_plain_controller.h#L60)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

long line
</blockquote></details>

Done.b

---

*[manipulation/schunk_wsg/test/schunk_wsg_controller_test.cc, line 33 at r1](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LCio2-512NheHZkHOH_:-LDJ3U5tDdjYfzjnxGVu:b58m5dw) ([raw file](https://github.com/robotlocomotion/drake/blob/e23ede3fe4cfc656a0fdf1a4db387500e10661b9/manipulation/schunk_wsg/test/schunk_wsg_controller_test.cc#L33)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

(apologies in advance, I think I'm missing something...)  In the previous case, we were only controlling one actuator, so only populating the state of the finger being directly actuated seemed sufficient.  Should this test now populate the state and look at the commanded force for two actuators?
</blockquote></details>

Good catch. I've updated the test.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDJ2GYbAf2bPo9xDnzQ:b-i9215u)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
@drake-jenkins-bot linux-xenial-clang-bazel-experimental-debug please

---

Review status: 4 of 16 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDMp51F97OExNXmGnB5:b-j9ui69)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
That timeout was real. The test wasn't respecting the `--simulation_sec` argument. That's fixed in the most recent revision.

---

Review status: 4 of 16 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDMvBdj5UV4LsS-o6is:bi77sm4)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
CI is actually happy now.

---

Review status: 4 of 16 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDNCfEy74Dq4RqDNk1n:b7bcf0)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> +@sherm1 for platform review per rotation

---

Reviewed 3 of 16 files at r1, 9 of 9 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDNPXTEEFY9Psf0zRJG:b-lkbyk7)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
+(status: do not merge) 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDNc8Au1t1rxCSRmhX1:b-91mz89)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
-(status: do not merge) 

---

Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDNciLGEb8QB-zrwn_x:b-b9fcpn)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Sweet! Much simpler than the old model. Platform <img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> with an assortment of BTWs (mostly doc).

---

Reviewed 7 of 16 files at r1, 9 of 9 files at r2.
Review status: all files reviewed at latest revision, all discussions resolved.

---

*[examples/kuka_iiwa_arm/iiwa_world/multi_arm_with_gripper_demo.cc, line 149 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDOCGxJ5Q-Er69K5u48:-LDOCGxJ5Q-Er69K5u49:b-thet2v) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/examples/kuka_iiwa_arm/iiwa_world/multi_arm_with_gripper_demo.cc#L149)):*
> ```cpp
>             max_force);
>     diagram_builder->Connect(max_force_source->get_output_port(),
>                     controller->get_input_port_max_force());
> ```

BTW indenting?

---

*[examples/kuka_iiwa_arm/iiwa_world/test/iiwa_wsg_diagram_factory_test.cc, line 15 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDOCVxz41MilT9dd8Vx:-LDOCVy-00AD4NKvE94s:bta7vvt) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/examples/kuka_iiwa_arm/iiwa_world/test/iiwa_wsg_diagram_factory_test.cc#L15)):*
> ```cpp
> namespace {
> 
> constexpr int kNumIiwaPositions = 7;
> ```

FYI consider adding a comment above these constants saying why these are the right numbers.

---

*[manipulation/schunk_wsg/schunk_wsg_constants.h, line 33 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO1H032fUugRW_H1wY:-LDO1H032fUugRW_H1wZ:bbr3m9p) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_constants.h#L33)):*
> ```objc
>  * Returns the position vector corresponding to the open position of the
>  * gripper. This is more complicated than one might expect due to the linkage in
>  * our model of the gripper.
> ```

BTW this comment is out of date now. Are there others that mention the linkage?

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.h, line 90 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO3CDP95vxYomKIFDl:-LDO3CDP95vxYomKIFDm:boipid) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.h#L90)):*
> ```objc
>  *                    └──────────┘
>  *```
>  * The remaining blocks differ only in their numerical parameters.
> ```

FYI beautiful documentation!

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.h, line 102 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO3VWE0l6nfjKGORIr:-LDO3VWE0l6nfjKGORIs:b25i830) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.h#L102)):*
> ```objc
>   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(SchunkWsgPlainController)
>   SchunkWsgPlainController(ControlMode control_mode = ControlMode::kPosition,
>                            double kp = 2000, double ki = 0, double kd = 5);
> ```

BTW needs `explicit`.
Also, seems awkward to have all these default parameters with the other signature having none. Consider these two as a more-straightforward alternative:
```c++
/** Specify control gains and mode. Mode defaults to position control. */
SchunkWsgPlainController(double kp, double ki, double kd,
                         ControlMode control_mode = ControlMode::kPosition);

/** Construct controller with default values for parameters and mode. 
    Gains are p/i/d=2000/0/5, mode is position control. */
SchunkWsgPlainController();
```

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.h, line 104 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO3I4M1ULn2Wa3ezoz:-LDO3I4M1ULn2Wa3ezp-:b7n8pfo) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.h#L104)):*
> ```objc
>                            double kp = 2000, double ki = 0, double kd = 5);
> 
>   SchunkWsgPlainController(double kp, double ki, double kd)
> ```

BTW needs a doxygen comment saying that this one defaults to position control.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.h, line 120 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO5eb-CXctFZZvnplW:-LDO5eb-CXctFZZvnplX:b-lzsex9) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.h#L120)):*
> ```objc
>   }
> 
>   // Implement StateFeedbackControllerInterface
> ```

BTW is this supposed to be a TODO?

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.cc, line 29 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO8u8rBl5JsLQ4FUSN:-LDO8u8rBl5JsLQ4FUSO:bdldlyl) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.cc#L29)):*
> ```cpp
>   constexpr int joint_state_size{2 * joint_position_size};
>   constexpr int mean_finger_position_size{1};
>   constexpr int grip_position_size{1};  // Distance between the fingers.
> ```

BTW not clear how a "distance" is a "size"? Is it the number of scalars used to represent the grip distance?

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.cc, line 40 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO6iBU3YEjHtExh3PH:-LDO6iBU3YEjHtExh3PI:bj82wzv) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.cc#L40)):*
> ```cpp
>   const auto mean_finger_position_jacobian =
>       (Matrix<double, mean_finger_position_size, joint_position_size>() <<
>         0.5, 0.5)
> ```

BTW can you say why these are the right values? Not obvious to me. 

FYI Would be nice to have some equations to look at. Side comment -- the long variable names are very descriptive when I first see them (good) but then are hard to interpret as math when they are used below (not so good). Consider whether it might be better to define shorter "math-y" symbols at the start, like
```
nq  Number of joint positions (two fingers).
nx  Size of state x (joint positions and velocities).
nd  Size of distance representation (one scalar) [?]
nm Size of mean separation representation (one scalar) [?]
Jg  Grip position Jacobian (nd x nq)
Jf  Finger position Jacobian (nm x nq)
etc.
```
Which would retain the good descriptive property but then would be easier to interpret when stuffed into blocks of matrices later.

Alternatively, consider comments before each matrix construction below showing the relevant math.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.cc, line 54 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO7B0W0hprtBewmjFJ:-LDO7B0W0hprtBewmjFK:b-5pdozy) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.cc#L54)):*
> ```cpp
>           (MatrixX<double>(control_position_size, joint_position_size)
>                << grip_position_jacobian,
>            mean_finger_position_jacobian)
> ```

BTW odd alignment

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.cc, line 87 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDOAD2o--ohNhmgmcBI:-LDOAD2o--ohNhmgmcBJ:b-j4i3j3) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.cc#L87)):*
> ```cpp
>           VectorX<double>::Constant(control_position_size, kp),
>           VectorX<double>::Constant(control_position_size, ki),
>           VectorX<double>::Constant(control_position_size, kd));
> ```

FYI, as an example of what I mean the above is IMO easier to read and understand at a glance than it would be if the variables were named `position_gain`, `integral_gain` etc. 

---

*[manipulation/schunk_wsg/test/schunk_wsg_controller_test.cc, line 20 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDOBGolCZ6qor3JKBRE:-LDOBGolCZ6qor3JKBRF:b-fvz4qq) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/test/schunk_wsg_controller_test.cc#L20)):*
> ```cpp
> /// Runs the controller for a brief period with the specified initial
> /// conditions and returns the commanded force.
> std::pair<double, double> RunWsgControllerTestStep(
> ```

BTW comment needs updating? Maybe just force -> forces and specify which is which?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDOD7Cv6NvJbL2Ie4D4:b-87wjbd)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Addressed platform reviewer comments. PTAL.

---

Review status: 8 of 18 files reviewed at latest revision, 5 unresolved discussions.

---

*[examples/kuka_iiwa_arm/iiwa_world/multi_arm_with_gripper_demo.cc, line 149 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDOCGxJ5Q-Er69K5u48:-LDgL7000zw9Lvb-acKr:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/examples/kuka_iiwa_arm/iiwa_world/multi_arm_with_gripper_demo.cc#L149)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW indenting?
</blockquote></details>

Done.

---

*[examples/kuka_iiwa_arm/iiwa_world/test/iiwa_wsg_diagram_factory_test.cc, line 15 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDOCVxz41MilT9dd8Vx:-LDgOOdo1tsmfg8Wz3sA:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/examples/kuka_iiwa_arm/iiwa_world/test/iiwa_wsg_diagram_factory_test.cc#L15)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

FYI consider adding a comment above these constants saying why these are the right numbers.
</blockquote></details>

Done.

---

*[manipulation/schunk_wsg/schunk_wsg_constants.h, line 33 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO1H032fUugRW_H1wY:-LDgShNbBqT2dDF-q6ZY:b-nb0bu3) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_constants.h#L33)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW this comment is out of date now. Are there others that mention the linkage?
</blockquote></details>

Good call. Grep found two more. 

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.h, line 102 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO3VWE0l6nfjKGORIr:-LDgXQlf4sLH8mfNPTvN:bijeld4) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.h#L102)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW needs `explicit`.
Also, seems awkward to have all these default parameters with the other signature having none. Consider these two as a more-straightforward alternative:
```c++
/** Specify control gains and mode. Mode defaults to position control. */
SchunkWsgPlainController(double kp, double ki, double kd,
                         ControlMode control_mode = ControlMode::kPosition);

/** Construct controller with default values for parameters and mode. 
    Gains are p/i/d=2000/0/5, mode is position control. */
SchunkWsgPlainController();
```


</blockquote></details>

For my edification, can you explain why it needs `expllicit`? Also, I've taken the even more straightforward course and deleted the second constructor.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.h, line 104 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO3I4M1ULn2Wa3ezoz:-LDgXlB96Dl8l62FkQsv:b-k7mx7q) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.h#L104)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW needs a doxygen comment saying that this one defaults to position control.
</blockquote></details>

Deleted.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.h, line 120 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO5eb-CXctFZZvnplW:-LDgVUk39MrIp0c-G-Jy:b-fki2ui) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.h#L120)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW is this supposed to be a TODO?
</blockquote></details>

No it's descriptive of the  methods that follow it. Hopefully that's clearer in the most recent revision.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.cc, line 29 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO8u8rBl5JsLQ4FUSN:-LDhRBB5ABR-kzUKqPwm:btfj3lj) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.cc#L29)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW not clear how a "distance" is a "size"? Is it the number of scalars used to represent the grip distance?
</blockquote></details>

Reworked per your comment below. Hopfully it's clearer now. PTAL.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.cc, line 40 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO6iBU3YEjHtExh3PH:-LDhRGBEAoTz80PvANtq:b-tn61hd) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.cc#L40)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW can you say why these are the right values? Not obvious to me. 

FYI Would be nice to have some equations to look at. Side comment -- the long variable names are very descriptive when I first see them (good) but then are hard to interpret as math when they are used below (not so good). Consider whether it might be better to define shorter "math-y" symbols at the start, like
```
nq  Number of joint positions (two fingers).
nx  Size of state x (joint positions and velocities).
nd  Size of distance representation (one scalar) [?]
nm Size of mean separation representation (one scalar) [?]
Jg  Grip position Jacobian (nd x nq)
Jf  Finger position Jacobian (nm x nq)
etc.
```
Which would retain the good descriptive property but then would be easier to interpret when stuffed into blocks of matrices later.

Alternatively, consider comments before each matrix construction below showing the relevant math.
</blockquote></details>

Went with your first suggestion. PTAL.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.cc, line 54 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO7B0W0hprtBewmjFJ:-LDgOwJV2khWSZbYM7gQ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.cc#L54)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW odd alignment
</blockquote></details>

Done.

---

*[manipulation/schunk_wsg/test/schunk_wsg_controller_test.cc, line 20 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDOBGolCZ6qor3JKBRE:-LDgTt_D17A6Qa48UgBQ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/test/schunk_wsg_controller_test.cc#L20)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW comment needs updating? Maybe just force -> forces and specify which is which?
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDhRMNXA4Bn8zv4E2Jn:b9qz1p5)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Looks great -- love the new notation! I noticed some substantive changes since last I looked at this -- please consider whether you want the feature reviewer to look them over pre-merge.

---

Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.h, line 102 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO3VWE0l6nfjKGORIr:-LDhXAXY7w0G-6_woJX_:b6nil7o) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.h#L102)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

For my edification, can you explain why it needs `expllicit`? Also, I've taken the even more straightforward course and deleted the second constructor.
</blockquote></details>

Because this method has defaults for all 4 parameters, it can be called with just one (the ControlMode). That makes it an implicit conversion from ControlMode to SchunkWsgPlainController unless you say `explicit` to prevent that.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.h, line 112 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDhXbkM9ZGKH7HmkrHu:-LDhXbkM9ZGKH7HmkrHv:b-rdyrfv) ([raw file](https://github.com/robotlocomotion/drake/blob/8626699aa05be1c97fab0ffdbd8836880d68c228/manipulation/schunk_wsg/schunk_wsg_plain_controller.h#L112)):*
> ```objc
>   /** Specify control gains and mode. Mode defaults to position control. */
>   SchunkWsgPlainController(ControlMode control_mode = ControlMode::kPosition,
>                            double kp = 2000, double ki = 0, double kd = 5);
> ```

FYI this is an awkward API because it it allows invocation with 0,1,2,3 or 4 parameters but not all of those combinations make sense. For example
`SchunkWsgPlainController(ControlMode::kVelocity)` still has a big position gain -- not obvious to me that the defaults you've listed make sense for velocity.

IMO a cleaner interface would be  to have a base constructor that requires all four parameters, then one that takes just the ControlMode and generates mode-dependent defaults, then invokes the base constructor.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.cc, line 40 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO6iBU3YEjHtExh3PH:-LDh_ZgKAQd18aOjsLHS:b-w0z4ti) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.cc#L40)):*
<details><summary><i>Previously, avalenzu (Andres Valenzuela) wrote…</i></summary><blockquote>

Went with your first suggestion. PTAL.
</blockquote></details>

Beautiful! Thanks.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.cc, line 61 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDhZhANAqTG1nAwO28h:-LDhZhANAqTG1nAwO28i:b96wn5j) ([raw file](https://github.com/robotlocomotion/drake/blob/8626699aa05be1c97fab0ffdbd8836880d68c228/manipulation/schunk_wsg/schunk_wsg_plain_controller.cc#L61)):*
> ```cpp
>   // controller state and the controller output to joint force.
> 
>   // Size of the control position, c. The control position is the quanity
> ```

BTW quanity -> quantity

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.cc, line 78 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDh_ew13A3PPF918mFb:-LDh_ew13A3PPF918mFc:b5mc0i1) ([raw file](https://github.com/robotlocomotion/drake/blob/8626699aa05be1c97fab0ffdbd8836880d68c228/manipulation/schunk_wsg/schunk_wsg_plain_controller.cc#L78)):*
> ```cpp
>     }
>   }
>   const int control_state_size{2 * nc};
> ```

FTI now `control_state_size` is conspicuously different than the nx notation you used for the other sizes (it's fine though). Still, consider inventing a compact notation (ns?) for this too for consistency.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDha9IA4C-otCJHfvgC:bx1bcst)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.h, line 102 at r2](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDO3VWE0l6nfjKGORIr:-LDlCDxr3W-A-Yc4TIiS:b9nf5v1) ([raw file](https://github.com/robotlocomotion/drake/blob/5d86f69b86c2d4b9e9368c0c6613f13662e55ab9/manipulation/schunk_wsg/schunk_wsg_plain_controller.h#L102)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

Because this method has defaults for all 4 parameters, it can be called with just one (the ControlMode). That makes it an implicit conversion from ControlMode to SchunkWsgPlainController unless you say `explicit` to prevent that.
</blockquote></details>

Got it. Thanks! Your last comment made this blocking, I believe.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.h, line 112 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDhXbkM9ZGKH7HmkrHu:-LDlBo8jChZIxv9DQICl:b-qhalcr) ([raw file](https://github.com/robotlocomotion/drake/blob/8626699aa05be1c97fab0ffdbd8836880d68c228/manipulation/schunk_wsg/schunk_wsg_plain_controller.h#L112)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

FYI this is an awkward API because it it allows invocation with 0,1,2,3 or 4 parameters but not all of those combinations make sense. For example
`SchunkWsgPlainController(ControlMode::kVelocity)` still has a big position gain -- not obvious to me that the defaults you've listed make sense for velocity.

IMO a cleaner interface would be  to have a base constructor that requires all four parameters, then one that takes just the ControlMode and generates mode-dependent defaults, then invokes the base constructor.
</blockquote></details>

There is no `ControlMode::kVelocity`. The PID gains are always applied to position-ish quantities, either the mean of the finger positions, or the difference between their positions.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.cc, line 61 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDhZhANAqTG1nAwO28h:-LDlEjQS2cVbnjHy4cPJ:b-896fix) ([raw file](https://github.com/robotlocomotion/drake/blob/8626699aa05be1c97fab0ffdbd8836880d68c228/manipulation/schunk_wsg/schunk_wsg_plain_controller.cc#L61)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

BTW quanity -> quantity
</blockquote></details>

Done.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.cc, line 78 at r3](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDh_ew13A3PPF918mFb:-LDlEq9q9-DsB1F8_-2Y:b951gjb) ([raw file](https://github.com/robotlocomotion/drake/blob/8626699aa05be1c97fab0ffdbd8836880d68c228/manipulation/schunk_wsg/schunk_wsg_plain_controller.cc#L78)):*
<details><summary><i>Previously, sherm1 (Michael Sherman) wrote…</i></summary><blockquote>

FTI now `control_state_size` is conspicuously different than the nx notation you used for the other sizes (it's fine though). Still, consider inventing a compact notation (ns?) for this too for consistency.
</blockquote></details>

Since this variable was only used once, I've removed it. Also, I think you have a typo that made this blocking.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819)*
<!-- Sent from Reviewable.io -->

sherm1 commented:
Sorry for inadvertent blocks -- I'm not sure I have the hang yet of the new discussion categories. Anyway, just one open comment now, ptal. I _think_ you can resolve it yourself after a fix.

---

Review status: all files reviewed at latest revision, all discussions resolved.

---

*[manipulation/schunk_wsg/schunk_wsg_plain_controller.h, line 111 at r4](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDlTQ1a-fHnj5Oq5fZg:-LDlTQ1a-fHnj5Oq5fZh:bw63jm2) ([raw file](https://github.com/robotlocomotion/drake/blob/b8f5610f0b27d5deed257b7153f99ef639525eec/manipulation/schunk_wsg/schunk_wsg_plain_controller.h#L111)):*
> ```objc
>   DRAKE_NO_COPY_NO_MOVE_NO_ASSIGN(SchunkWsgPlainController)
>   /** Specify control gains and mode. Mode defaults to position control. */
>   SchunkWsgPlainController(ControlMode control_mode = ControlMode::kPosition,
> ```

This still needs `explicit`.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDlTjL46o92VQ-LibRr:bpfo891)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Thanks, @sherm1!

---

Review status: 17 of 18 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDlYW4i6eohf73aUmPq:bbxcqhp)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
+(status: do not merge) while I investigate new CI failure.

---

Review status: 17 of 18 files reviewed at latest revision, all discussions resolved, some commit checks failed.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDlsGGZ7FACPuy2dIek:b-s247hs)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
-(status: do not merge)  decreasing the integrator step size slightly resolved the failing test cases locally.

---

Review status: 16 of 18 files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDlwbLJ8hV_Fi1Oj21t:b-e7ja9p)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> (again)  Still works well in the simulation I've been testing with and delivers a performance win (maybe not quite 2x anymore, but I think that's because of other changes to my test case).

---

Review status: all files reviewed at latest revision, all discussions resolved.

---

*[manipulation/schunk_wsg/schunk_wsg_lcm.cc, line 135 at r5](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDmZmYh8TYXoYN_EWsw:-LDmZmYiEMtIAzruVlyX:b-er492s) ([raw file](https://github.com/robotlocomotion/drake/blob/ed4c3c073a5362ed5b96ab5ee02e2390d90018b5/manipulation/schunk_wsg/schunk_wsg_lcm.cc#L135)):*
> ```cpp
> 
>   drake::log()->info("cur_position = {}", cur_position);
>   drake::log()->info("target_position = {}", target_position);
> ```

leftover debug prints?

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDmeDxd-KSeDGHtek8Q:blavl6z)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
Review status: 17 of 18 files reviewed at latest revision, 1 unresolved discussion.

---

*[manipulation/schunk_wsg/schunk_wsg_lcm.cc, line 135 at r5](https://reviewable.io/reviews/robotlocomotion/drake/8819#-LDmZmYh8TYXoYN_EWsw:-LDmehhI0nduG58XFZEq:b-59gw5z) ([raw file](https://github.com/robotlocomotion/drake/blob/ed4c3c073a5362ed5b96ab5ee02e2390d90018b5/manipulation/schunk_wsg/schunk_wsg_lcm.cc#L135)):*
<details><summary><i>Previously, sammy-tri (Sam Creasey) wrote…</i></summary><blockquote>

leftover debug prints?
</blockquote></details>

Sigh. Yup!  Removed.

---


*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819)*
<!-- Sent from Reviewable.io -->

avalenzu commented:
-(status: do not merge) 

---

Review status: 17 of 18 files reviewed at latest revision, 1 unresolved discussion.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819#-:-LDqLbJF6X3u5-RWWTKi:b-b9fcpn)*
<!-- Sent from Reviewable.io -->

sammy-tri commented:
Review status: all files reviewed at latest revision, all discussions resolved.

---



*Comments from [Reviewable](https://reviewable.io/reviews/robotlocomotion/drake/8819)*
<!-- Sent from Reviewable.io -->
